builder/dockerfile/internals.go

```diff
@@ -0,0 +1,657 @@
+
+	// Copy the ONBUILD triggers, and remove them from the config, since the config will be committed.
+	onBuildTriggers := b.runConfig.OnBuild
+	b.runConfig.OnBuild = []string{}
+
+	// parse the ONBUILD triggers by invoking the parser
+	for _, step := range onBuildTriggers {
+		ast, err := parser.Parse(strings.NewReader(step))
+		if err != nil {
+			return err
+		}
+
+		for i, n := range ast.Children {
+			switch strings.ToUpper(n.Value) {
+			case "ONBUILD":
+				return fmt.Errorf("Chaining ONBUILD via `ONBUILD ONBUILD` isn't allowed")
+			case "MAINTAINER", "FROM":
+				return fmt.Errorf("%s isn't allowed as an ONBUILD trigger", n.Value)
+			}
+
+			if err := b.dispatch(i, n); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// probeCache checks if `b.docker` implements builder.ImageCache and image-caching
+// is enabled (`b.UseCache`).
+// If so attempts to look up the current `b.image` and `b.runConfig` pair with `b.docker`.
+// If an image is found, probeCache returns `(true, nil)`.
+// If no image is found, it returns `(false, nil)`.
+// If there is any error, it returns `(false, err)`.
+func (b *Builder) probeCache() (bool, error) {
+	c, ok := b.docker.(builder.ImageCache)
+	if !ok || !b.UseCache || b.cacheBusted {
+		return false, nil
+	}
+	cache, err := c.GetCachedImage(builder.ImageID(b.image), b.runConfig)
+	if err != nil {
+		return false, err
+	}
+	if len(cache) == 0 {
+		logrus.Debugf("[BUILDER] Cache miss")
+		b.cacheBusted = true
+		return false, nil
+	}
+
+	fmt.Fprintf(b.Stdout, " ---> Using cache\n")
+	logrus.Debugf("[BUILDER] Use cached version")
+	b.image = string(cache)
+
+	// TODO: remove once Commit can take a tag parameter.
+	b.docker.Retain(b.id, b.image)
+	b.activeImages = append(b.activeImages, b.image)
+
+	return true, nil
+}
+
+func (b *Builder) create() (*daemon.Container, error) {
+	if b.image == "" && !b.noBaseImage {
+		return nil, fmt.Errorf("Please provide a source image with `from` prior to run")
+	}
+	b.runConfig.Image = b.image
+
+	// TODO: why not embed a hostconfig in builder?
+	hostConfig := &runconfig.HostConfig{
+		CPUShares:    b.CPUShares,
+		CPUPeriod:    b.CPUPeriod,
+		CPUQuota:     b.CPUQuota,
+		CpusetCpus:   b.CPUSetCpus,
+		CpusetMems:   b.CPUSetMems,
+		CgroupParent: b.CgroupParent,
+		Memory:       b.Memory,
+		MemorySwap:   b.MemorySwap,
+		Ulimits:      b.Ulimits,
+	}
+
+	config := *b.runConfig
+
+	// Create the container
+	c, warnings, err := b.docker.Create(b.runConfig, hostConfig)
+	if err != nil {
+		return nil, err
+	}
+	defer c.Unmount()
+	for _, warning := range warnings {
+		fmt.Fprintf(b.Stdout, " ---> [Warning] %s\n", warning)
+	}
+
+	b.tmpContainers[c.ID] = struct{}{}
+	fmt.Fprintf(b.Stdout, " ---> Running in %s\n", stringid.TruncateID(c.ID))
+
+	if config.Cmd.Len() > 0 {
+		// override the entry point that may have been picked up from the base image
+		s := config.Cmd.Slice()
+		c.Path = s[0]
+		c.Args = s[1:]
+	} else {
+		config.Cmd = stringutils.NewStrSlice()
```

Comment:
I'm not sure what this mean, config is a copy and it doesn't seem to left this func.